src/lib.rs

```diff
@@ -32,6 +32,52 @@ use core::{fmt, ops::DerefMut};
 pub mod block;
 pub mod le;
 
+/// A random generator
+pub trait Generator {
+    /// The result type.
+    ///
+    /// This could be a simple word like `u64` or an array like `[u32; 16]`.
+    type Result;
+
+    /// Generate a new result.
+    ///
+    /// Since [`Self::Result`] may be large, the output is passed by reference.
+    /// Word generators should likely implement this as a shim over another
+    /// method:
+    /// ```
+    /// pub struct CountingGenerator(u64);
+    /// impl CountingGenerator {
+    ///     fn next(&mut self) -> u64 {
+    ///         let x = self.0;
+    ///         self.0 = self.0.wrapping_add(1);
+    ///         x
+    ///     }
+    /// }
+    ///
+    /// impl rand_core::Generator for CountingGenerator {
+    ///     type Result = u64;
+    ///
+    ///     #[inline]
+    ///     fn generate(&mut self, result: &mut Self::Result) {
+    ///         *result = self.next();
+    ///     }
+    /// }
+    /// ```
+    fn generate(&mut self, result: &mut Self::Result);
```

Comment:
Generating by-reference has two limitations:

- Using it requires first constructing a "default" instance. Since we don't even require `Result: Default`, this is only usable when the `Result` type is known or otherwise bounded. In practice this seems fine for our use-cases, but it *could* be an issue.
- Using it to generate small, `Copy` types is unergonomic and dependent on inlining for good performance.

Alternative 1: return `Self::Result` by value. This is easier to use. For `chacha20` (uses 2048-bit output blocks) it may be dependent on inlining optimisations for performance. (This type of optimisation is commonly relied on by constructors in Rust.)

Alternative 2: have both fns. Over-complex.